compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala

```diff
@@ -682,16 +682,304 @@ abstract class GenJSCode extends plugins.PluginComponent
+        overloadIdent: js.Ident, dispatchResolution: js.Tree,
+        isConstructor: Set[String])(implicit pos: Position): List[js.Tree] = {
+      if (!constructorGraph.hasSubConstructors) {
+        dispatchResolution match {
+          case js.Block(stats) =>
+            val js.ApplyStatic(_, method, _) = stats.last
+            val refs = constructorGraph.getParamRefs(method.name)
+            val rhss = stats.collect { case js.VarDef(_, _, _, rhs) => rhs }
+            zipMap(refs, rhss) { (ref, rhs) =>
+              js.VarDef(ref.ident, ref.tpe, false, rhs)
+            }
+
+          case js.ApplyStatic(cls, method, args) if isConstructor(method.name) =>
+            // Calls the constructor with no arguments
+            Nil
+
+          case tree => List(tree)
+        }
+      } else {
+        val overloadRef = js.VarRef(overloadIdent)(jstpe.IntType)
+        def assignOverloadNumbers(tree: js.Tree): js.Tree = tree match {
+          case js.Block(stats) =>
+            val js.ApplyStatic(_, method, _) = stats.last
+            val num = constructorGraph.getOverrideNum(method.name)
+            val refs = overloadRef :: constructorGraph.getParamRefs(method.name)
+            val rhss = js.IntLiteral(num) ::
+              stats.collect { case js.VarDef(_, _, _, rhs) => rhs }
+            val newStats = zipMap(refs, rhss)((ref, rhs) => js.Assign(ref, rhs))
+            js.Block(newStats)
+
+          case js.ApplyStatic(cls, method, args) if isConstructor(method.name) =>
+            // Calls the constructor with no arguments
+            js.Assign(overloadRef,
+                js.IntLiteral(constructorGraph.getOverrideNum(method.name)))
+
+          case js.Match(selector, cases, default) =>
+            val newCases = cases.map {
+              case (literals, body) => (literals, assignOverloadNumbers(body))
+            }
+            js.Match(selector, newCases, default)(tree.tpe)
+
+          case js.If(cond, thenp, elsep) =>
+            js.If(cond, assignOverloadNumbers(thenp),
+              assignOverloadNumbers(elsep))(tree.tpe)
+
+          case _ => tree
+        }
+
+        val newDispatchResolution = assignOverloadNumbers(dispatchResolution)
+        val allParamDefsAsVars = constructorGraph.getAllParamDefsAsVars
+        val overrideNumDef =
+          js.VarDef(overloadIdent, jstpe.IntType, true, js.IntLiteral(0))
+
+        overrideNumDef :: allParamDefsAsVars ::: newDispatchResolution :: Nil
+      }
+    }
+
+    private def mkConstructorGraph(primaryCtor: js.MethodDef,
+        secondaryCtors: List[js.MethodDef], isConstructor: Set[String])(
+        implicit pos: Position): ConstructorGraph = {
+      def superCall(tree: js.Tree): Option[String] = tree match {
+        case js.Block(stats) =>
+          stats.map(superCall).collectFirst {
+            case Some(name) => name
+          }
+
+        case js.ApplyStatic(cls, method, args) if isConstructor(method.name) =>
+          Some(method.name)
```

Comment:
You should be able to use `ir.Definitions.isConstructorName(method.name)` instead. You should probably also check that `args.head` is a `This()`, and that `cls` is `currentClassType`.